src/Lykke.AlgoStore.MatchingEngineAdapter.Services/MatchingEngineAdapter.cs

```diff
@@ -0,0 +1,217 @@
+            {MeStatusCodes.ReservedVolumeHigherThanBalance, ResponseModel.ErrorCodeType.ReservedVolumeHigherThanBalance},
+            {MeStatusCodes.NotFound, ResponseModel.ErrorCodeType.NotFound},
+            {MeStatusCodes.BalanceLowerThanReserved, ResponseModel.ErrorCodeType.BalanceLowerThanReserved},
+            {MeStatusCodes.LeadToNegativeSpread, ResponseModel.ErrorCodeType.LeadToNegativeSpread},
+            {MeStatusCodes.TooSmallVolume, ResponseModel.ErrorCodeType.Dust},
+            {MeStatusCodes.InvalidFee, ResponseModel.ErrorCodeType.Runtime},
+            {MeStatusCodes.Duplicate, ResponseModel.ErrorCodeType.Runtime},
+            {MeStatusCodes.Runtime, ResponseModel.ErrorCodeType.Runtime}
+        };
+
+        public MatchingEngineAdapter(IMatchingEngineClient matchingEngineClient,
+            IAlgoInstanceTradeRepository algoClientInstanceRepository,
+            IFeeCalculatorClient feeCalculatorClient,
+            IAssetsService assetsService,
+            string feeSettingsTargetClientIdHft,
+            [NotNull] ILog log)
+        {
+            _matchingEngineClient =
+                matchingEngineClient ?? throw new ArgumentNullException(nameof(matchingEngineClient));
+            _algoInstanceTradeRepository = algoClientInstanceRepository ?? throw new ArgumentNullException(nameof(algoClientInstanceRepository));
+            _feeCalculatorClient = feeCalculatorClient ?? throw new ArgumentNullException(nameof(feeCalculatorClient));
+            _assetsService = assetsService ?? throw new ArgumentNullException(nameof(assetsService));
+            _feeSettingsTargetClientIdHft = feeSettingsTargetClientIdHft ?? throw new ArgumentNullException(nameof(_feeSettingsTargetClientIdHft));
+            _log = log ?? throw new ArgumentNullException(nameof(log));
+        }
+
+        public async Task<ResponseModel> CancelLimitOrderAsync(Guid limitOrderId)
+        {
+            var response = await _matchingEngineClient.CancelLimitOrderAsync(limitOrderId.ToString());
+            await CheckResponseAndThrowIfNull(response);
+
+            return ConvertToApiModel(response.Status);
+        }
+
+        public async Task<ResponseModel<double>> HandleMarketOrderAsync(string clientId, string assetPairId, OrderAction orderAction, double volume,
+            bool straight, string instanceId, double? reservedLimitVolume = null)
+        {
+            var order = new MarketOrderModel
+            {
+                Id = GetNextRequestId().ToString(),
+                AssetPairId = assetPairId,
+                ClientId = clientId,
+                ReservedLimitVolume = reservedLimitVolume,
+                Straight = straight,
+                Volume = Math.Abs(volume),
+                OrderAction = orderAction.ToMeOrderAction(),
+                Fees = new[] { await GetMarketOrderFee(clientId, assetPairId, orderAction) }
+            };
+
+            var response = await _matchingEngineClient.HandleMarketOrderAsync(order);
```

Comment:
Send a cancellation token. Like here https://github.com/LykkeCity/Lykke.Service.FixGateway/blob/dev/src/Lykke.Service.FixGateway.Services/NewOrderRequestHandler.cs#L187
It protects you from hanging on this method in certain circumstances. It is true for all ME API methods

Reply:
We were told to use the HFT code as an example and use its code, there is no such token, is this a new requirement?
https://github.com/LykkeCity/Lykke.Service.HFT/blob/master/src/Lykke.Service.HFT.Services/MatchingEngineAdapter.cs#L85